craft_application/commands/lifecycle.py

```diff
@@ -128,11 +137,35 @@ def run(
     ) -> None:
         """Run a lifecycle step command."""
         super().run(parsed_args)
+
+        shell = getattr(parsed_args, "shell", False)
+        shell_after = getattr(parsed_args, "shell_after", False)
+        debug = getattr(parsed_args, "debug", False)
+
         step_name = step_name or self.name
-        self._services.lifecycle.run(
-            step_name=step_name,
-            part_names=parsed_args.parts,
-        )
+
+        if shell:
```

Comment:
what prevents or controls the inside/outside nature here? will we get one shell inside and then another outside?

Reply:
No; the command is only run once, by the application running inside the instance.

When in regular mode, the "outside" application launches the "inside one" and only the inside one actually runs this command code. See the flow [here](https://github.com/canonical/craft-application/blob/main/craft_application/application.py#L256).